### DISCUSSION TITLE
¿cómo calculo la complejidad de esta función?

### DISCUSSION THREAD

**Discussion by IoachimusRoderici** [INGCOM-UNRN-P1/cursada-2023#212] [Preguntas y respuestas]
Estoy con el ejercicio de la suma de dígitos de Fibonacci y no sé como calcular el costo de la función.
El código es este:

```C
int suma_fibonacci (int n){
    if (n < 2){                     // 2 operaciones
        return 0;                   // 1
    }

    int i = 2;                      // 1
    int F_i = 1;                    // 1
    int F_i_1 = 0;                  // 1
    int total = 1;                  // 1
    int aux;

    for ( ; i < n; i++){            // 3 en la condición, 2 en el cuerpo
        aux = F_i + F_i_1;          // 4
        F_i_1 = F_i;                // 2
        F_i = aux;                  // 2
        
        total += suma_digitos(F_i); // 3 + (10·log10(F_i) + 14) + 1
    }

    return total;
}
```

El costo de la función `suma_digitos(n)` ya lo calculé y me da $10·log_{10}(n) + 14$. Pero como la llamada a esta función está adentro del bucle y su costo cambia en cada repetición, no sé como hacer la suma de todas las repeticiones.

**Comment by TeoInMaghine** [comment]
Creo que se apunta al peor caso, entonces sería $[10·log_{10}(n) + 14]n$ (se multiplicaría con $n$ el resto de costos que hay en el for también).

**Reply by IoachimusRoderici** [reply]
Pero pasa que el argumento de `suma_digitos` no es `i` sino `F_i` (el *i*ésimo número de fibonacci), que crece mucho más rápido que `i`.

**Comment by IoachimusRoderici** [comment]
Posible respuesta:

1. Los números de Fibonacci crecen exponencialmente y están acotados por $F_i \leq 2^i$ ([fuente](https://math.stackexchange.com/questions/894743/proof-by-induction-nth-fibonacci-number-is-at-most-2n)).
2. El costo de calcular `suma_digitos(F_i)` es a lo sumo el costo de `suma_digitos(2^i)`.
3. La complejidad de `suma_digitos(n)` es $O(log(n))$, así que la de `suma_digitos(2^i)` es $O(\log 2^i) = O(i)$.
4. El bucle tiene `n` repeticiones y la complejidad de cada una es a lo sumo $O(n)$, así que la complejidad del bucle es $O(n^2)$.

¿Algún adulto responsable puede confirmar esto?

**Comment by martinvilu** [comment]
El código enviado es de baja legibilidad, los nombres de las variables no ayudan, así como su organización que no favorece el cálculo de la complejidad.

El cálculo del n-esimo término de Fibonacci es una cosa, con su propio costo y la suma de los dígitos de un número es otro.

El cálculo es más complejo de lo que debiera de ser.

**Reply by IoachimusRoderici** [reply]
¿Así está más legible?

**Consigna:**
> TP2 - Ej. 9: Escribe una función que calcule la suma de los dígitos de los términos en una serie de Fibonacci hasta el N-ésimo término.

```C
/* Recibe un número n y devuelve la suma de los
 * dígitos de los primeros n términos en la
 * secuencia de fibonacci.
 * Pre: n >= 0
 * Pos: suma_fibonacci(n) >= 0
 */
int suma_fibonacci (int n){
   //Los primeros dos casos dan 0:
   if (n < 2){
      return 0;
   }

   //Para n >= 2:
   int i = 2;
   int F_i = 1;   //el i-ésimo número de Fibonacci
   int F_i_1 = 0; //el (i-1)-ésimo número de Fibonacci
   int total = 1;
   int aux;

   for ( ; i < n; i++){
      //Calcular el siguiente término:
      aux = F_i + F_i_1;
      F_i_1 = F_i;
      F_i = aux;

      //Sumar los dígitos:
      total += suma_digitos(F_i);
   }

   return total;
}
```

**Reply by IoachimusRoderici** [reply]
> su organización que no favorece el cálculo de la complejidad.

¿Cómo se podría cambiar la organización para facilitar el cálculo?

> El cálculo del n-esimo término de Fibonacci es una cosa, con su propio costo y la suma de los dígitos de un número es otro.

> El cálculo es más complejo de lo que debiera de ser.

No entiendo cómo esto es más complejo de lo que debería ser, ¿me podrías sugerir una manera menos compleja?

De todos modos, la pregunta era cómo calcular la complejidad de esta función, no si la complejidad era la mejor posible.

**Reply by martinvilu** [reply]
> ¿Cómo se podría cambiar la organización para facilitar el cálculo?

Fibonacci en una función, suma de dígitos en otra; unidos en la función que llama a estas dos.

El cálculo es más complejo porque hay superpuesta la obtención del término de Fibonacci con la llamada a la función de suma de dígitos. La organización de código simplifica la obtención de estos valores.

**Reply by IoachimusRoderici** [reply]
Organizar el código de esa manera haría que sea más fácil de analizar, pero también haría que sea mucho más costoso.

Algo por el estilo de 

```C
for (int i = 2; i<n; i++){
    fibonacci_i = fibonacci(i);
    total += suma_digitos(fibonacci_i);
}
```

Tendría que volver a calcular toda la serie de fibonacci en cada vuelta del bucle; en cambio mi código reusa los términos que se habían calculado en el paso anterior.

Esto es más "limpio" pero es mucho menos eficiente. Yo creo que es mejor la opción más eficiente, aunque sea un poquito más difícil de analizar.

**Reply by martinvilu** [reply]
> Organizar el código de esa manera haría que sea más fácil de analizar, pero también haría que sea mucho más costoso.

Explique su maqueta joven, ¿cómo cambian los valores de complejidad?

> es mucho menos eficiente

No deja de ser una optimización temprana sin razón, en particular porque no hemos visto como el compilador cambia el programa y probablemente resuelva por su cuenta esta ‘ineficiencia’ percibida.

**Reply by IoachimusRoderici** [reply]
> probablemente resuelva por su cuenta esta ‘ineficiencia’ percibida.

¿El compilador es suficientemente inteligente como para juntar dos bucles que están en funciones distintas en uno solo?

Si es así, retiro lo dicho.

**Reply by martinvilu** [reply]
Investígalo y contame. 

De todas formas, optimizar ciegamente no es una buena idea, y en especial, cuando no hace falta (y hace del código más complejo de lo necesario)

Siempre podés reemplazar la función de Fibonacci por una versión O(1)

**Comment by martinvilu** [comment]
```C
int suma_fibonacci (int n){
    if (n < 2){                     // 2 operaciones
        return 0;                   // 1
    }

    int i = 2;                      // 1
    int F_i = 1;                    // 1
    int F_i_1 = 0;                  // 1
    int total = 1;                  // 1
    int aux;

    for ( ; i < n; i++){            // 3 en la condición, 2 en el cuerpo
        aux = F_i + F_i_1;          // 4
        F_i_1 = F_i;                // 2
        F_i = aux;                  // 2
        
        total += suma_digitos(F_i); // 3 + (10·log10(F_i) + 14) + 1
    }

    return total;
}
```

Los nombres de las variables van en minúsculas, y múltiples nombres no son descriptivos; en particular, `F_i`, `F_i_1`; ¿qué es? La letra `i` sola, se utiliza para contadores, que en este caso no lo son.
